File: zenoh_pc_nodes/python/multiTopicSub.py
#!/usr/bin/env python3
import sys
import os

# Ensure the shared zenoh_ros package can be imported
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig, z_std_msgs

class MultiTopicSubscriberNode(ZenohNode):
    def __init__(self) -> None:
        super().__init__("multi_topic_subscriber")
        print(f"[Node] {self.z_get_name()} has been started")

        # Create subscription with standard ROS 2 message type (Int32) and depth 10
        self.counter_sub = self.z_create_subscription(
            z_std_msgs.Int32,
            "robot/sim_counter",
            self.counter_listener_callback,
            10
        )

        # Create subscription with standard ROS 2 message type (String) and depth 10
        self.string_sub = self.z_create_subscription(
            z_std_msgs.String,
            "robot/hello_string",
            self.string_listener_callback,
            10
        )

    def counter_listener_callback(self, msg: z_std_msgs.Int32) -> None:
        print(f"[RECV COUNTER] sim_counter: {msg.data}")

    def string_listener_callback(self, msg: z_std_msgs.String) -> None:
        print(f"[RECV STRING] hello_string: {msg.data}")


def main() -> None:
    # Initialize with default AP network settings (IP: 192.168.4.1, Port: 7447)
    config = ZenohConfig(host="192.168.4.1", port=7447)
    ZenohNode.init(config)
    node_instance = None
    try:
        node_instance = MultiTopicSubscriberNode()
        node_instance.z_spin()

    except KeyboardInterrupt:
        print("\n[CTRL+C]>>> Interrupted by the User.")
        
    except Exception as e:
        print(f"Critical Error: {e}")
        
    finally:
        if node_instance is not None:
            print("Destroying the Zenoh Node...")
            node_instance.z_destroy()


if __name__ == '__main__':
    main()

File: zenoh_pc_nodes/cpp/src/multi_topic_sub.cpp
#include <zenoh_ros/ZenohRosPC.h>
#include <zenoh_ros/std_msgs/z_Int32.h>
#include <zenoh_ros/std_msgs/z_String.h>

class MultiTopicSubscriberNode : public ZenohNode {
public:
    MultiTopicSubscriberNode() : ZenohNode("multi_topic_subscriber") {
        ZLOG_INFO(this->get_logger(), "Node has been started");

        // Create subscription to robot/sim_counter (Int32)
        counter_sub_ = this->z_create_subscription<z_Int32>(
            "robot/sim_counter",
            [this](const z_Int32& msg) -> void {
                this->counter_listener_callback(msg);
            },
            10
        );

        // Create subscription to robot/hello_string (String)
        string_sub_ = this->z_create_subscription<z_String>(
            "robot/hello_string",
            [this](const z_String& msg) -> void {
                this->string_listener_callback(msg);
            },
            10
        );
    }

private:
    ZenohSubscription<z_Int32>* counter_sub_ = nullptr;
    ZenohSubscription<z_String>* string_sub_ = nullptr;

    void counter_listener_callback(const z_Int32& msg) {
        ZLOG_INFO(this->get_logger(), "[RECV COUNTER] sim_counter: %d", msg.data);
    }

    void string_listener_callback(const z_String& msg) {
        ZLOG_INFO(this->get_logger(), "[RECV STRING] hello_string: %s", msg.data.c_str());
    }
};

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    // Initialize with default AP network settings (IP: 192.168.4.1, Port: 7447)
    ZenohConfig config;
    config.host = "192.168.4.1";
    config.port = 7447;

    if (!ZenohNode::init(config)) {
        return -1;
    }

    {
        MultiTopicSubscriberNode node;
        node.z_spin();
    } // Node goes out of scope and destroys subscriptions cleanly before session closes via atexit

    return 0;
}
